op-node/node/node.go

```diff
@@ -132,136 +152,214 @@ func New(ctx context.Context, cfg *config.Config, log log.Logger, appVersion str
-	n.initEventSystem()
-	if err := n.initL1BeaconAPI(ctx, cfg); err != nil {
-		return err
+	if overrides == nil {
+		overrides = &initializationOverrides{}
 	}
-	if err := n.initL1Source(ctx, cfg); err != nil {
-		return fmt.Errorf("failed to init L1 Source: %w", err)
+
+	eventSys, eventDrain, err := initEventSystem(n)
+	if err != nil {
+		return fmt.Errorf("failed to init event system: %w", err)
+	}
+	n.eventSys = eventSys
+	n.eventDrain = eventDrain
+
+	if overrides.Beacon == nil {
+		beacon, err := initL1BeaconAPI(ctx, cfg, n)
+		if err != nil {
+			return err
+		}
+		n.beacon = beacon
+	} else {
+		n.beacon = overrides.Beacon
+	}
+
+	if overrides.L1Source == nil {
+		l1Source, err := initL1Source(ctx, cfg, n)
+		if err != nil {
+			return fmt.Errorf("failed to init L1 Source: %w", err)
+		}
+		n.l1Source = l1Source
+	} else {
+		n.l1Source = overrides.L1Source
 	}
-	if err := n.initL2(ctx, cfg); err != nil {
+
+	// initL2 may use side effects to register interop subsystem to the node.EventSystem
+	l2Source, interopSys, l2Driver, safeDB, err := initL2(ctx, cfg, n)
+	if err != nil {
 		return fmt.Errorf("failed to init L2: %w", err)
 	}
-	if err := n.initL1Handlers(cfg); err != nil {
-		return fmt.Errorf("failed to init L1 Handlers: %w", err)
+	n.l2Source = l2Source
+	n.interopSys = interopSys
+	n.l2Driver = l2Driver
+	n.safeDB = safeDB
+
+	l1HeadsSub, l1SafeSub, l1FinalizedSub, err := initL1Handlers(cfg, n)
+	if err != nil {
+		return fmt.Errorf("failed to init L1 Source: %w", err)
 	}
-	if err := n.initRuntimeConfig(ctx, cfg); err != nil { // depends on L2, to signal initial runtime values to
+	n.l1HeadsSub = l1HeadsSub
+	n.l1SafeSub = l1SafeSub
+	n.l1FinalizedSub = l1FinalizedSub
+
+	// initRuntimeConfig relies on side effects to set the runCfg, node.halted and call node.cancel if needed
+	if err := initRuntimeConfig(ctx, cfg, n); err != nil {
 		return fmt.Errorf("failed to init the runtime config: %w", err)
 	}
-	if err := n.initP2PSigner(ctx, cfg); err != nil {
+
+	p2pSigner, err := initP2PSigner(ctx, cfg, n)
+	if err != nil {
 		return fmt.Errorf("failed to init the P2P signer: %w", err)
 	}
-	if err := n.initP2P(cfg); err != nil {
+	n.p2pSigner = p2pSigner
+
+	p2pNode, err := initP2P(cfg, n)
+	if err != nil {
 		return fmt.Errorf("failed to init the P2P stack: %w", err)
 	}
+	n.p2pNode = p2pNode
+
 	// Only expose the server at the end, ensuring all RPC backend components are initialized.
-	if err := n.initRPCServer(cfg); err != nil {
-		return fmt.Errorf("failed to init the RPC server: %w", err)
+	if overrides.RPCHandler == nil {
+		server, err := initRPCServer(cfg, n)
+		if err != nil {
+			return fmt.Errorf("failed to init the RPC server: %w", err)
+		}
+		n.server = server
+	} else {
+		// the node registers to an existing RPC server's handler if provided
+		// the node assumes the RPC server is already started
+		n.server = nil
+		err := registerAPIs(cfg, n, overrides.RPCHandler)
```

Comment:
nit: I wonder if we should make `initRPCServer` return the RPCHandler and then we could use the same flow to registerAPIs in the override and non-override case. Otherwise we have two separate paths to register APIs and they could potentially diverge (the code is shared through registerAPIs but it's possible some important handler setup could get added to initRPCServer directly).